src/hotspot/share/memory/allocation.hpp

```diff
@@ -263,6 +263,24 @@ class ClassLoaderData;
 class MetaspaceClosure;
 
 class MetaspaceObj {
+  // This class has no virtual functions (except for Metadata and its
+  // subtypes). Nevertheless, there are some functions that all subtypes of
+  // MetaspaceObj are expected to implement, so that templates which are
+  // defined for this class hierarchy can work uniformly.
+  //
+  // The following functions are required by MetaspaceClosure:
+  //   void metaspace_pointers_do(MetaspaceClosure* it) { <walk my refs> }
+  //   int size() const { return align_up(sizeof(<This>), wordSize) / wordSize; }
+  //   MetaspaceObj::Type type() const { return <This>Type; }
+  //
+  // The following functions are required by MetadataFactory::free_metadata():
+  //   bool on_stack() { return false; } // DEBUG only
```

Comment:
I don't think the // DEBUG only comment is correct.  on_stack is used in product mode.  Can you remove this?

Reply:
Hmm .... in product code, there's no use of `MetaspaceObj::on_stack()`, but there's use of `Metadata::on_stack()`.